### ISSUE TITLE
AdamWeightDecay seems not a general implement.

### ISSUE THREAD

**Issue by qiuxin2012** [intel-analytics/analytics-zoo#2408]
Seems AdamWeightDecay is not a general Adam implement with weightdecay, it's much different from pytorch's [Adam](https://pytorch.org/docs/stable/_modules/torch/optim/adam.html#Adam).
Pytorch's decay comes first, and has bias_correction.

**Comment by jason-dai** [comment]
Are you using this one https://github.com/intel-analytics/analytics-zoo/blob/master/zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/optimizers/Adam.scala? I think it's similar to Keras

**Comment by jason-dai** [comment]
Or this one https://github.com/intel-analytics/analytics-zoo/blob/master/zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/optimizers/AdamWeightDecay.scala? It's based on the TensorFlow BERT implementation @dding3 ?

**Comment by qiuxin2012** [comment]
> Or this one https://github.com/intel-analytics/analytics-zoo/blob/master/zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/optimizers/AdamWeightDecay.scala? It's based on the TensorFlow BERT implementation @dding3 ?

This is the one I was talking about, it's much different from pytorch's adam.

**Comment by qiuxin2012** [comment]
> Are you using this one https://github.com/intel-analytics/analytics-zoo/blob/master/zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/optimizers/Adam.scala? I think it's similar to Keras

Yes, this one doesn't have weight decay. In SPDB's case, they are using torch.optim.Adam(model.parameters(), lr=learning_rate, weight_decay=1e-5).

**Comment by dding3** [comment]
https://github.com/intel-analytics/analytics-zoo/blob/master/zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/optimizers/AdamWeightDecay.scala is based on the Adam Proposed in BERT paper.
